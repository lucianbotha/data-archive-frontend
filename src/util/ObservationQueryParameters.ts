// TypeScript interfaces describing the state of the search form

/**
 * An interface describing the React state of the observation query form.
 *
 * This state is used for populating all search form fields (and error messages)
 * and is parsed into the GraphQL query for searching observations.
 *
 * Properties:
 * -----------
 * general:
 *     General information about the observation.
 * target:
 *     Target details.
 * telescope:
 *     Telescope (and instrument) details.
 */
export interface IObservationQueryParameters {
  general?: IGeneral;
  target?: ITarget;
  telescope: ITelescope;
}

/**
 * An interface for query parameters related to general information.
 *
 * Properties:
 * -----------
 * errors:
 *     Error messages.
 * observationNight:
 *     Date of the night when the observation was taken. This
 *     refers to the date whe the night started.
 * proposalCode:
 *     Identifier of the proposal to which the observation belongs.
 * proposalTitle:
 *     Title of the proposal to which the observation belongs.
 *
 */
export interface IGeneral {
  errors: {
    observationNight?: string;
  };
  observationNight?: string;
  principalInvestigator?: string;
  proposalCode?: string;
}

/**
 * An interface for query parameters related to target details.
 *
 * Properties:
 * -----------
 * declination:
 *     Declination.
 * errors:
 *     Errors.
 * name:
 *     Target name.
 * resolver:
 *     Resolver to use for resolving the target name to a position.
 * rightAscension:
 *     Right ascension.
 * searchConeRadius:
 *     Radius of the search cone radius, in the units given by
 *     searchConeRadiusUnits.
 * searchConeRadiusUnits:
 *     Units of the search cone radius given by searchConeRadius.
 *
 */
export interface ITarget {
  declination?: string;
  errors: {
    declination?: string;
    rightAscension?: string;
    searchConeRadius?: string;
    searchConeRadiusUnits?: string;
  };
  name?: string;
  resolver?: string;
  rightAscension?: string;
  searchConeRadius?: string;
  searchConeRadiusUnits?: string;
}

// TELESCOPES

export type TelescopeName = "Lesedi" | "SALT" | "1.9 m";

/**
 * An interface for query parameters related to a telescope.
 *
 * This serves as a placeholder for more concrete interfaces such as ISALT.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface ITelescope {
  name: TelescopeName;
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 */
export interface ISALT extends ITelescope {
  name: "SALT";
  instrument?: IInstrument;
}

/**
 * An interface for query parameters related to Lesedi.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 */
export interface ILesedi extends ITelescope {
  name: "Lesedi";
  instrument?: IInstrument;
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 */
export interface ILesedi extends ITelescope {
  name: "Lesedi";
  instrument?: IInstrument;
}

// INSTRUMENTS

/**
 * An interface for query parameters related to an instrument.
 *
 * This serves as a placeholder for more concrete interfaces such as ISalticam.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface IInstrument {
  name: string;
}

/**
 * An interface for query parameters related to Salticam.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface ISalticam extends IInstrument {
  name: "Salticam";
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}

/**
 * An interface for query parameters related to RSS.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface IRSS extends IInstrument {
  name: "RSS";
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}

/**
 * An interface for query parameters related to HRS.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * mode:
 *     Resolution mode.
 */
export interface IHRS extends IInstrument {
  name: "HRS";
  errors: {
    exposureTime?: string;
    mode?: string;
  };
  exposureTime?: string;
  mode?:
    | "Low Resolution"
    | "Medium Resolution"
    | "High Resolution"
    | "High Stability";
}

/**
 * An interface for query parameters related to HIPPO.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface IHIPPO extends IInstrument {
  name: "HIPPO";
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}

/**
 * An interface for query parameters related to SHOC.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 */
export interface ISHOC extends IInstrument {
  name: "SHOC";
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
}
