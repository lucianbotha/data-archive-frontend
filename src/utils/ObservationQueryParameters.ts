// TypeScript interfaces describing the state of the search form

/**
 * An interface describing the React state of the observation query form.
 *
 * This state is used for populating all search form fields (and error messages)
 * and is parsed into the GraphQL query for searching observations.
 *
 * Properties:
 * -----------
 * general:
 *     General information about the observation.
 * target:
 *     Target details.
 * telescope:
 *     Telescope (and instrument) details.
 */
export interface IObservationQueryParameters {
  general: IGeneral;
  target: ITarget;
  telescope: ITelescope;
}

/**
 * An interface for query parameters related to general information.
 *
 * Properties:
 * -----------
 * errors:
 *     Error messages.
 * observationNight:
 *     Date of the night when the observation was taken. This
 *     refers to the date whe the night started.
 * proposalCode:
 *     Identifier of the proposal to which the observation belongs.
 * proposalTitle:
 *     Title of the proposal to which the observation belongs.
 *
 */
export interface IGeneral {
  errors: {
    observationNight?: string;
    principalInvestigator?: string;
    proposalCode?: string;
    proposalTitle?: string;
    arcs?: string;
    biases?: string;
    flats?: string;
    standards?: string;
    dataType?: string;
  };
  observationNight?: string;
  principalInvestigator?: string;
  proposalCode?: string;
  proposalTitle?: string;
  arcs?: boolean;
  biases?: boolean;
  flats?: boolean;
  standards?: boolean;
  dataType?: "any" | "reduced" | "raw";
}

/**
 * An interface for query parameters related to target details.
 *
 * Properties:
 * -----------
 * declination:
 *     Declination.
 * errors:
 *     Errors.
 * name:
 *     Target name.
 * resolver:
 *     Resolver to use for resolving the target name to a position.
 * rightAscension:
 *     Right ascension.
 * searchConeRadius:
 *     Radius of the search cone radius, in the units given by
 *     searchConeRadiusUnits.
 * searchConeRadiusUnits:
 *     Units of the search cone radius given by searchConeRadius.
 *
 */
export interface ITarget {
  declination?: string;
  loading?: string;
  errors: {
    name?: string;
    declination?: string;
    resolver?: string;
    rightAscension?: string;
    searchConeRadius?: string;
    searchConeRadiusUnits?: string;
  };
  name?: string;
  resolver?: "Simbad" | "NED" | "VizieR";
  rightAscension?: string;
  searchConeRadius?: string;
  searchConeRadiusUnits?: string;
}

// TELESCOPES

/**
 * An interface for query parameters related to a telescope.
 *
 * This serves as a placeholder for more concrete interfaces such as ISALT.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface ITelescope {
  name?: "Lesedi" | "SALT" | "1.9 m";
}

/**
 * An interface for query parameters related to SALT.
 *
 * Properties:
 * -----------
 * instrument:
 *     A SALT instrument.
 * name:
 *     The string "SALT".
 */
export interface ISALT extends ITelescope {
  instrument?: IInstrument;
  name: "SALT";
}

/**
 * An interface for query parameters related to Lesedi.
 *
 * Properties:
 * -----------
 * instrument:
 *     An instrument used on Lesedi.
 * name:
 *     The string "Lesedi".
 */
export interface ILesedi extends ITelescope {
  instrument?: IInstrument;
  name: "Lesedi";
}

/**
 * An interface for query parameters related to the 1.9 m Telescope.
 *
 * Properties:
 * -----------
 * instrument:
 *     An instrument used on the 1./9 m Telescope.
 * name:
 *     The string "1.9 m".
 */
export interface I1Dot9Metre extends ITelescope {
  instrument?: IInstrument;
  name: "1.9 m";
}

// INSTRUMENTS

/**
 * An interface for query parameters related to an instrument.
 *
 * This serves as a placeholder for more concrete interfaces such as ISalticam.
 *
 * Properties:
 * -----------
 * name:
 *     Telescope name.
 */
export interface IInstrument {
  name: string;
}

/**
 * An interface for query parameters related to Salticam.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "Salticam".
 */
export interface ISalticam extends IInstrument {
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "Salticam";
}

/**
 * An interface for query parameters related to RSS.
 *
 * Properties:
 * -----------
 * detectorMode:
 *     Detector mode.
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "RSS".
 */
export interface IRSS extends IInstrument {
  detectorMode?: "Normal" | "Slot Mode";
  errors: {
    detectorMode?: string;
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "RSS";
}

/**
 * An interface for query parameters related to HRS.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * mode:
 *     Resolution mode.
 * name:
 *     The string "HRS".
 */
export interface IHRS extends IInstrument {
  errors: {
    exposureTime?: string;
    mode?: string;
  };
  exposureTime?: string;
  mode?:
    | "Low Resolution"
    | "Medium Resolution"
    | "High Resolution"
    | "High Stability";
  name: "HRS";
}

/**
 * An interface for query parameters related to HIPPO.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "HIPPO".
 */
export interface IHIPPO extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "HIPPO";
}

/**
 * An interface for query parameters related to SHOC.
 *
 * Properties:
 * -----------
 * errors:
 *     Errors.
 * exposureTime:
 *     Exposure time.
 * filter:
 *     Filter.
 * name:
 *     The string "SHOC".
 */
export interface ISHOC extends IInstrument {
  errors: {
    exposureTime?: string;
    filter?: string;
  };
  exposureTime?: string;
  filter?: string;
  name: "SHOC";
}

export interface ISearchResults {
  results?: {
    name: string | "Unknown Observation";
    files: {
      filename: string;
      name: string;
      dataType: string;
      isRedused: boolean;
      targetName: string;
      rightAscension: string;
      declination: string;
      observationNight: string;
      category: string;
      telescope: string;
      instrument: string;
      proposalCode: string;
    }[];
  }[];
}
