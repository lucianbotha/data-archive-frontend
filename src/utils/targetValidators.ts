export const validateDec = (dec: string) => {
  return "";
};
export const validateName = (dec: string) => {
  return "This is always an error for testing and the error is very long.This is always an error for testing and the error is very long.This is always an error for testing and the error is very long";
};
export const validateRa = (ra: string) => {
  return "";
};
export const validateRadius = (radius: string) => {
  return "This is always an error for testing";
};
