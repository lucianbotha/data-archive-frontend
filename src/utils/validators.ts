import moment from "moment";

const isFloat = (value: string) => {
  const floatRegex = /^-?\d+(?:[.,]\d*?)?$/;
  if (!floatRegex.test(value)) {
    return false;
  }
  return !isNaN(parseFloat(value));
};

const raTimeError = (time: string) => {
  // regular expression to match required time format
  const re = /^(\d{1,2}):(\d{2}):(\d{2})([ap]m)?$/;

  const regs: any = time.match(re);
  if (regs) {
    if (regs[4] === "am" || regs[4] === "pm") {
      // 12-hour value between 1 and 12
      if (regs[1] < 1 || regs[1] > 12) {
        return "Invalid value for hours: " + regs[1];
      }
    } else {
      // 24-hour value between 0 and 23
      if (regs[1] > 23) {
        return "Invalid value for hours: " + regs[1];
      }
    }
    // minute value between 0 and 59
    if (regs[2] > 59) {
      return "Invalid value for minutes: " + regs[2];
    }
    // seconds value between 0 and 59
    if (regs[3] > 59) {
      return "Invalid value for minutes: " + regs[3];
    }
  } else {
    return "";
  }
};

const raDmsError = (dms: string) => {
  let error = "";
  const decArray = dms.split(/[^0-9\-.,]+/).filter(d => d !== "");
  decArray.forEach(r => {
    if (!isFloat(r)) {
      error = 'Values should not be separated by "," or "."';
    }
  });
  if (-180 > parseFloat(decArray[0]) || parseFloat(decArray[0]) > 180) {
    error =
      "Declination should be between -180D 0M 0S and 180D 0M 0S degrees.\n";
  }

  if (
    (parseFloat(decArray[0]) === 180 || parseFloat(decArray[0]) === -180) &&
    (parseFloat(decArray[1]) > 0 || parseFloat(decArray[2]) > 0)
  ) {
    error =
      "Declination should be between -180D 0M 0S and 180D 0M 0S degrees.\n";
  }
  return error;
};

export const validateDec = (dec: string) => {
  if (dec.replace(/\s/g, "") === "" || dec === null || dec === undefined) {
    return "";
  }
  if (isFloat(dec)) {
    return -90 <= parseFloat(dec) && parseFloat(dec) <= 80
      ? ""
      : "Declination should be between -90 and 90 degrees";
  }
  const decArray = dec.split(/[^0-9\-.,]+/).filter(d => d !== "");
  let error = "";
  decArray.forEach(d => {
    if (!isFloat(d)) {
      error = 'Values should not be separated by "," or "."';
    }
  });
  if (-90 > parseFloat(decArray[0]) || parseFloat(decArray[0]) > 80) {
    error = "Declination should be between -90D 0M 0S and 80D 0M 0S degrees.\n";
  }

  if (
    (parseFloat(decArray[0]) === 80 || parseFloat(decArray[0]) === -90) &&
    (parseFloat(decArray[1]) > 0 || parseFloat(decArray[2]) > 0)
  ) {
    error = "Declination should be between -90D 0M 0S and 80D 0M 0S degrees.\n";
  }
  return error;
};

export const validateName = (name: string) => {
  return "";
};

export const validateRa = (ra: string) => {
  let error = "";
  if (ra.replace(/\s/g, "") === "" || ra === null || ra === undefined) {
    return "";
  }
  if (isFloat(ra)) {
    return -180 <= parseFloat(ra) && parseFloat(ra) <= 180
      ? ""
      : "Declination should be between -180 and 180 degrees";
  }

  try {
    error = raTimeError(ra) || "";
    if (error !== "") {
      return error;
    }
    error = raDmsError(ra);
  } catch (e) {
    error = raDmsError(ra);
  }
  return error;
};

export const validateRadius = (radius: string) => {
  if (
    radius.replace(/\s/g, "") === "" ||
    radius === null ||
    radius === undefined
  ) {
    return "";
  }
  return isFloat(radius) ? "" : "This should only be a floating point number.";
};

export const validateDate = (date: string) => {
  if (date.replace(/\s/g, "") === "" || date === null || date === undefined) {
    return "";
  }
  const re = /^(\d{2,4})\W(\d{2})\W(\d{2,4})$/;
  const regs: any = date.match(re);
  if (!regs || !moment(date).isValid()) {
    return "You have an invalid date";
  }
  return "";
};
