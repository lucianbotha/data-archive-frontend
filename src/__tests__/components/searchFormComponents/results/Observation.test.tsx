import { mount, shallow } from "enzyme";
import toJson from "enzyme-to-json";
import * as React from "react";
import ObservationResults from "../../../../components/searchFormComponents/results/Observation";
import { IObservation } from "../../../../utils/ObservationQueryParameters";
const addAllFiles = jest.fn();
describe("Observation results table row", () => {
  it("should render", () => {
    expect(
      mount(
        <ObservationResults
          cart={[]}
          observation={{ files: [] } as any}
          addAllFiles={addAllFiles}
        />
      )
    ).toBeDefined();
  });
  it("should always display a row if observation is given", () => {
    const observation: IObservation = {
      files: [],
      id: "Obs-ID-xxx",
      name: "observation",
      proposal: "P-CODE-XXX",
      startTime: "2019-10-10 23:00:12",
      telescope: "SALT"
    };
    const wrapper = shallow(
      <ObservationResults
        cart={[]}
        observation={observation}
        addAllFiles={addAllFiles}
      />
    );
    expect(wrapper.find("tr").length).toEqual(1);
  });
  it("should have a checkbox to add all files if there is an observation", () => {
    const observation: IObservation = {
      files: [
        {
          category: "some info",
          dataType: "some info",
          declination: "some info",
          filename: "name-XXX",
          instrument: "some info",
          isReduced: true,
          name: "some info",
          observationId: "some info",
          observationNight: "some info",
          rightAscension: "some info",
          targetName: "some info",
          telescope: "some info",
          url: "some info"
        }
      ],
      id: "Obs-ID-xxx",
      name: "observation",
      proposal: "P-CODE-XXX",
      startTime: "2019-10-10 23:00:12",
      telescope: "SALT"
    };
    const cart = [
      {
        category: "some info",
        dataType: "some info",
        declination: "some info",
        filename: "name-XXX",
        instrument: "some info",
        isReduced: true,
        name: "some info",
        observationId: "some info",
        observationNight: "some info",
        rightAscension: "some info",
        targetName: "some info",
        telescope: "some info",
        url: "some info"
      }
    ];
    const wrapper = mount(
      <ObservationResults
        cart={cart}
        observation={observation}
        addAllFiles={addAllFiles}
      />
    );

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
