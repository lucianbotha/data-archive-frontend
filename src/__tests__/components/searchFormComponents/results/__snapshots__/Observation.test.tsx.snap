// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Observation results table row should have a checkbox to add all files if there is an observation 1`] = `
.c0.c0 {
  font-size: 22px;
  padding-right: 10px;
}

.c1.c1 {
  width: 18px;
  height: 18px;
}

<ObservationResults
  addAllFiles={[MockFunction]}
  cart={
    Array [
      Object {
        "category": "some info",
        "dataType": "some info",
        "declination": "some info",
        "filename": "name-XXX",
        "instrument": "some info",
        "isReduced": true,
        "name": "some info",
        "observationId": "some info",
        "observationNight": "some info",
        "rightAscension": "some info",
        "targetName": "some info",
        "telescope": "some info",
        "url": "some info",
      },
    ]
  }
  observation={
    Object {
      "files": Array [
        Object {
          "category": "some info",
          "dataType": "some info",
          "declination": "some info",
          "filename": "name-XXX",
          "instrument": "some info",
          "isReduced": true,
          "name": "some info",
          "observationId": "some info",
          "observationNight": "some info",
          "rightAscension": "some info",
          "targetName": "some info",
          "telescope": "some info",
          "url": "some info",
        },
      ],
      "id": "Obs-ID-xxx",
      "name": "observation",
      "proposal": "P-CODE-XXX",
      "startTime": "2019-10-10 23:00:12",
      "telescope": "SALT",
    }
  }
>
  <tr
    className="is-selected span"
  >
    <td>
      <label>
        <styled.span>
          <StyledComponent
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [
                  Object {
                    "className": "span",
                  },
                ],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-iwsKbI",
                  "isStatic": true,
                  "lastClassName": "c0",
                  "rules": Array [
                    "
  && {
    font-size: 22px;
    padding-right: 10px;
  }
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-iwsKbI",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="span c0"
            >
              <styled.input
                checked={false}
                id="Add-all-Obs-ID-xxx"
                onChange={[Function]}
              >
                <StyledComponent
                  checked={false}
                  forwardedComponent={
                    Object {
                      "$$typeof": Symbol(react.forward_ref),
                      "attrs": Array [
                        Object {
                          "className": "checkbox",
                          "type": "checkbox",
                        },
                      ],
                      "componentStyle": ComponentStyle {
                        "componentId": "sc-gZMcBi",
                        "isStatic": true,
                        "lastClassName": "c1",
                        "rules": Array [
                          "
  && {
    width: 18px;
    height: 18px;
  }
",
                        ],
                      },
                      "displayName": "styled.input",
                      "foldedComponentIds": Array [],
                      "render": [Function],
                      "styledComponentId": "sc-gZMcBi",
                      "target": "input",
                      "toString": [Function],
                      "warnTooManyClasses": [Function],
                      "withComponent": [Function],
                    }
                  }
                  forwardedRef={null}
                  id="Add-all-Obs-ID-xxx"
                  onChange={[Function]}
                >
                  <input
                    checked={false}
                    className="checkbox c1"
                    id="Add-all-Obs-ID-xxx"
                    onChange={[Function]}
                    type="checkbox"
                  />
                </StyledComponent>
              </styled.input>
            </span>
          </StyledComponent>
        </styled.span>
        <styled.span
          className="span"
        >
          <StyledComponent
            className="span"
            forwardedComponent={
              Object {
                "$$typeof": Symbol(react.forward_ref),
                "attrs": Array [
                  Object {
                    "className": "span",
                  },
                ],
                "componentStyle": ComponentStyle {
                  "componentId": "sc-iwsKbI",
                  "isStatic": true,
                  "lastClassName": "c0",
                  "rules": Array [
                    "
  && {
    font-size: 22px;
    padding-right: 10px;
  }
",
                  ],
                },
                "displayName": "styled.span",
                "foldedComponentIds": Array [],
                "render": [Function],
                "styledComponentId": "sc-iwsKbI",
                "target": "span",
                "toString": [Function],
                "warnTooManyClasses": [Function],
                "withComponent": [Function],
              }
            }
            forwardedRef={null}
          >
            <span
              className="span span c0"
            >
               Add all
            </span>
          </StyledComponent>
        </styled.span>
      </label>
    </td>
    <td
      colSpan={3}
    >
      <styled.span>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [
                Object {
                  "className": "span",
                },
              ],
              "componentStyle": ComponentStyle {
                "componentId": "sc-iwsKbI",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  && {
    font-size: 22px;
    padding-right: 10px;
  }
",
                ],
              },
              "displayName": "styled.span",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-iwsKbI",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="span c0"
          >
            Observation: 
            observation
          </span>
        </StyledComponent>
      </styled.span>
    </td>
    <td
      colSpan={2}
    >
      <styled.span>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [
                Object {
                  "className": "span",
                },
              ],
              "componentStyle": ComponentStyle {
                "componentId": "sc-iwsKbI",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  && {
    font-size: 22px;
    padding-right: 10px;
  }
",
                ],
              },
              "displayName": "styled.span",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-iwsKbI",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="span c0"
          >
            Telescope: 
            SALT
          </span>
        </StyledComponent>
      </styled.span>
    </td>
    <td
      colSpan={2}
    >
      <styled.span>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [
                Object {
                  "className": "span",
                },
              ],
              "componentStyle": ComponentStyle {
                "componentId": "sc-iwsKbI",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  && {
    font-size: 22px;
    padding-right: 10px;
  }
",
                ],
              },
              "displayName": "styled.span",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-iwsKbI",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="span c0"
          >
            Proposal: 
            P-CODE-XXX
          </span>
        </StyledComponent>
      </styled.span>
    </td>
    <td
      colSpan={2}
    >
      <styled.span>
        <StyledComponent
          forwardedComponent={
            Object {
              "$$typeof": Symbol(react.forward_ref),
              "attrs": Array [
                Object {
                  "className": "span",
                },
              ],
              "componentStyle": ComponentStyle {
                "componentId": "sc-iwsKbI",
                "isStatic": true,
                "lastClassName": "c0",
                "rules": Array [
                  "
  && {
    font-size: 22px;
    padding-right: 10px;
  }
",
                ],
              },
              "displayName": "styled.span",
              "foldedComponentIds": Array [],
              "render": [Function],
              "styledComponentId": "sc-iwsKbI",
              "target": "span",
              "toString": [Function],
              "warnTooManyClasses": [Function],
              "withComponent": [Function],
            }
          }
          forwardedRef={null}
        >
          <span
            className="span c0"
          >
            Stat time: 
            2019-10-10 23:00:12
          </span>
        </StyledComponent>
      </styled.span>
    </td>
  </tr>
</ObservationResults>
`;
